{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "34cee08c-b6ee-450d-ae33-edd181df02b4",
   "metadata": {},
   "source": [
    "# Analyze the counts of the different variants\n",
    "Here we analyze the counts of different variants.\n",
    "\n",
    "First, import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a2b4fb9f-ad03-4348-8c99-1dfda78b0dde",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pickle\n",
    "\n",
    "import altair as alt\n",
    "\n",
    "import Bio.SeqIO\n",
    "\n",
    "import dms_variants.codonvarianttable\n",
    "import dms_variants.utils\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dfdbecf5-ead3-4d3e-a0ae-106cc124bce5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# allow more rows for Altair\n",
    "_ = alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "686a8cb3-92b6-4b41-aae4-e1d97907c8fd",
   "metadata": {},
   "source": [
    "Change working directory to top directory of repo:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4a4cbbc8-9cd4-46f7-923e-d32242183a36",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "os.chdir('../')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "649e0dcc-7e23-4386-84d8-3cafba0ec69b",
   "metadata": {},
   "source": [
    "Read configuration:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6aa10c89-ae3b-43da-a707-a6bef3d454d6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open('config.yaml') as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5874b79d-a130-4aa4-9917-60c33e0cac9a",
   "metadata": {},
   "source": [
    "Read the barcode counts and barcode runs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cbf2aa94-05c1-4669-baf4-190d0e1c9c15",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "barcode_runs = (\n",
    "    pd.read_csv(config[\"barcode_runs\"])\n",
    "    .assign(antibody=lambda x: x[\"antibody\"].fillna(\"no antibody\"))\n",
    ")\n",
    "\n",
    "barcode_counts = pd.read_csv(config[\"barcode_counts\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a291ee61-0679-4de8-855a-91341e4502ec",
   "metadata": {},
   "source": [
    "Read the parental gene sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0c4aa6e6-9033-49be-8048-e230db81fca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "geneseq = str(Bio.SeqIO.read(config['gene_sequence_codon'], 'fasta').seq)\n",
    "assert len(geneseq) % 3 == 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aae813f8-776c-46a4-b8ba-b24687149f1c",
   "metadata": {},
   "source": [
    "Get the primary target:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "521db153-3710-41dc-8df0-b11eaf5e895c",
   "metadata": {},
   "outputs": [],
   "source": [
    "primary_target = config[\"gene\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "290c93b5-60a2-45de-bd42-d3afc39ef4f1",
   "metadata": {},
   "source": [
    "## Create codon-variant table and add counts:\n",
    "Create a [dms_variants codon-variant table](https://jbloomlab.github.io/dms_variants/codonvariant_sim_data.html):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c98f4106-46ca-4de6-974e-a578bda6a0b8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "variants = dms_variants.codonvarianttable.CodonVariantTable(\n",
    "    barcode_variant_file=config[\"virus_variants_w_neut_standard\"],\n",
    "    geneseq=geneseq,\n",
    "    substitutions_are_codon=True,\n",
    "    substitutions_col=\"codon_substitutions\",\n",
    "    allowgaps=True,\n",
    "    primary_target=config[\"gene\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beea645e-8813-4295-b74e-45102070e746",
   "metadata": {},
   "source": [
    "Now add the sample counts:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a6b4bb48-6f7f-429f-a89f-ec81147f9db5",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants.add_sample_counts_df(barcode_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bad0374-3887-4e42-8895-a26621eb5d1c",
   "metadata": {},
   "source": [
    "Save the pickled codon-variant table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2ed404b7-34b5-43e3-a422-89dc9fe5d675",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(config[\"codon_variant_table_pickle\"], \"wb\") as f:\n",
    "    pickle.dump(variants, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b15dd9e-9e3d-4bbf-b0c2-db7e0a46dd3d",
   "metadata": {},
   "source": [
    "## Average mutations per variant\n",
    "Compute the average mutations per variant:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "930da3ba-946c-4404-95a0-8913b0bcbd9d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "avg_muts = (\n",
    "    variants.numCodonMutsByType(\n",
    "        variant_type=\"all\",\n",
    "        libraries=variants.libraries,\n",
    "    )\n",
    "    .merge(barcode_runs, validate=\"many_to_one\")\n",
    "    .drop(columns=[\"num_muts_count\", \"antibody_concentration\", \"replicate\", \"fastq_R1\", \"notes\"])\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d540952",
   "metadata": {},
   "source": [
    "Edit dataframe to only include data for talk, with clearer labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "b9c89ada",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>sample</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>count</th>\n",
       "      <th>number</th>\n",
       "      <th>date</th>\n",
       "      <th>virus_batch</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>antibody</th>\n",
       "      <th>library_sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>libA</td>\n",
       "      <td>220121_1_no-antibody_control_100x_1</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>7789129</td>\n",
       "      <td>2.741342</td>\n",
       "      <td>220121</td>\n",
       "      <td>1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "      <td>no antibody</td>\n",
       "      <td>libA_220121_1_no-antibody_control_100x_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>libA</td>\n",
       "      <td>220121_1_no-antibody_control_100x_1</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>7789129</td>\n",
       "      <td>0.172177</td>\n",
       "      <td>220121</td>\n",
       "      <td>1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "      <td>no antibody</td>\n",
       "      <td>libA_220121_1_no-antibody_control_100x_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>libA</td>\n",
       "      <td>220121_1_no-antibody_control_100x_1</td>\n",
       "      <td>stop</td>\n",
       "      <td>7789129</td>\n",
       "      <td>0.010285</td>\n",
       "      <td>220121</td>\n",
       "      <td>1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "      <td>no antibody</td>\n",
       "      <td>libA_220121_1_no-antibody_control_100x_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>libA</td>\n",
       "      <td>220121_1_no-antibody_control_100x_1</td>\n",
       "      <td>deletion</td>\n",
       "      <td>7789129</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>220121</td>\n",
       "      <td>1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "      <td>no antibody</td>\n",
       "      <td>libA_220121_1_no-antibody_control_100x_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>libA</td>\n",
       "      <td>220121_1_no-antibody_control_100x_2_rt1</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>5869174</td>\n",
       "      <td>2.738756</td>\n",
       "      <td>220121</td>\n",
       "      <td>1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "      <td>no antibody</td>\n",
       "      <td>libA_220121_1_no-antibody_control_100x_2_rt1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  library                                   sample  mutation_type    count  \\\n",
       "0    libA      220121_1_no-antibody_control_100x_1  nonsynonymous  7789129   \n",
       "1    libA      220121_1_no-antibody_control_100x_1     synonymous  7789129   \n",
       "2    libA      220121_1_no-antibody_control_100x_1           stop  7789129   \n",
       "3    libA      220121_1_no-antibody_control_100x_1       deletion  7789129   \n",
       "4    libA  220121_1_no-antibody_control_100x_2_rt1  nonsynonymous  5869174   \n",
       "\n",
       "     number    date  virus_batch          sample_type     antibody  \\\n",
       "0  2.741342  220121            1  no-antibody_control  no antibody   \n",
       "1  0.172177  220121            1  no-antibody_control  no antibody   \n",
       "2  0.010285  220121            1  no-antibody_control  no antibody   \n",
       "3  0.000034  220121            1  no-antibody_control  no antibody   \n",
       "4  2.738756  220121            1  no-antibody_control  no antibody   \n",
       "\n",
       "                                 library_sample  \n",
       "0      libA_220121_1_no-antibody_control_100x_1  \n",
       "1      libA_220121_1_no-antibody_control_100x_1  \n",
       "2      libA_220121_1_no-antibody_control_100x_1  \n",
       "3      libA_220121_1_no-antibody_control_100x_1  \n",
       "4  libA_220121_1_no-antibody_control_100x_2_rt1  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "avg_muts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "699ba7b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_61814/2329032379.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    }
   ],
   "source": [
    "# edit names for virus meeting, drop to just 220224 samples\n",
    "\n",
    "# avg_muts_reformatted = avg_muts.duplicate()\n",
    "\n",
    "short_names = {\n",
    "#     '220121_1_no-antibody_control_50x_1': 'no-Ab_1',\n",
    "#     '220121_1_no-antibody_control_50x_2_rt1': 'no-Ab_2_RT-1',\n",
    "#     '220121_1_no-antibody_control_50x_2_rt2': 'no-Ab_2_RT-2',\n",
    "    '220224_1_antibody_AUSAB-07_0.000158_1_rt1': 'serum-IC65.00',\n",
    "#     '220224_1_antibody_AUSAB-07_0.000158_1_rt2': 'serum-IC65_1_RT-2',\n",
    "#     '220224_1_antibody_AUSAB-07_0.000158_2': 'serum-IC65_2',\n",
    "    '220224_1_antibody_AUSAB-07_0.000395_1': 'serum-IC90.00',\n",
    "#     '220224_1_antibody_AUSAB-07_0.000395_2': 'serum-IC90_2',\n",
    "    '220224_1_antibody_AUSAB-07_0.00158_1': 'serum-IC99.00',\n",
    "#     '220224_1_antibody_AUSAB-07_0.00158_2': 'serum-IC99_2',\n",
    "    '220224_1_antibody_AUSAB-07_0.00632_1': 'serum-IC99.90',\n",
    "#     '220224_1_antibody_AUSAB-07_0.00632_2': 'serum-IC99.90_2',\n",
    "    '220224_1_antibody_AUSAB-07_0.0158_1': 'serum-IC99.99',\n",
    "#     '220224_1_antibody_AUSAB-07_0.0158_2_rt1': 'serum-IC99.99_2_RT-1',\n",
    "#     '220224_1_antibody_AUSAB-07_0.0158_2_rt2': 'serum-IC99.99_2_RT-2',\n",
    "    '220224_1_no-antibody_control_670ng_1': 'no-Ab',\n",
    "#     '220224_1_no-antibody_control_670ng_2': 'no-Ab_2',\n",
    "    '220224_1_no-antibody_control_plasmid': 'plasmid'\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "avg_muts['sample'] = avg_muts['sample'].map(short_names)\n",
    "avg_muts_reformatted = avg_muts.dropna(subset=['sample'])\n",
    "\n",
    "avg_muts_reformatted['condition'] = avg_muts_reformatted['sample'].str.split('_').str[0]                     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "6c9a4503",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>sample</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>count</th>\n",
       "      <th>number</th>\n",
       "      <th>date</th>\n",
       "      <th>virus_batch</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>antibody</th>\n",
       "      <th>library_sample</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>libA</td>\n",
       "      <td>serum-IC65.00</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>1892557</td>\n",
       "      <td>2.722493</td>\n",
       "      <td>220224</td>\n",
       "      <td>1</td>\n",
       "      <td>antibody</td>\n",
       "      <td>AUSAB-07</td>\n",
       "      <td>libA_220224_1_antibody_AUSAB-07_0.000158_1_rt1</td>\n",
       "      <td>serum-IC65.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>libA</td>\n",
       "      <td>serum-IC65.00</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>1892557</td>\n",
       "      <td>0.168752</td>\n",
       "      <td>220224</td>\n",
       "      <td>1</td>\n",
       "      <td>antibody</td>\n",
       "      <td>AUSAB-07</td>\n",
       "      <td>libA_220224_1_antibody_AUSAB-07_0.000158_1_rt1</td>\n",
       "      <td>serum-IC65.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>libA</td>\n",
       "      <td>serum-IC65.00</td>\n",
       "      <td>stop</td>\n",
       "      <td>1892557</td>\n",
       "      <td>0.010324</td>\n",
       "      <td>220224</td>\n",
       "      <td>1</td>\n",
       "      <td>antibody</td>\n",
       "      <td>AUSAB-07</td>\n",
       "      <td>libA_220224_1_antibody_AUSAB-07_0.000158_1_rt1</td>\n",
       "      <td>serum-IC65.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>libA</td>\n",
       "      <td>serum-IC65.00</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1892557</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>220224</td>\n",
       "      <td>1</td>\n",
       "      <td>antibody</td>\n",
       "      <td>AUSAB-07</td>\n",
       "      <td>libA_220224_1_antibody_AUSAB-07_0.000158_1_rt1</td>\n",
       "      <td>serum-IC65.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>libA</td>\n",
       "      <td>serum-IC90.00</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>2552542</td>\n",
       "      <td>2.712283</td>\n",
       "      <td>220224</td>\n",
       "      <td>1</td>\n",
       "      <td>antibody</td>\n",
       "      <td>AUSAB-07</td>\n",
       "      <td>libA_220224_1_antibody_AUSAB-07_0.000395_1</td>\n",
       "      <td>serum-IC90.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   library         sample  mutation_type    count    number    date  \\\n",
       "48    libA  serum-IC65.00  nonsynonymous  1892557  2.722493  220224   \n",
       "49    libA  serum-IC65.00     synonymous  1892557  0.168752  220224   \n",
       "50    libA  serum-IC65.00           stop  1892557  0.010324  220224   \n",
       "51    libA  serum-IC65.00       deletion  1892557  0.000031  220224   \n",
       "60    libA  serum-IC90.00  nonsynonymous  2552542  2.712283  220224   \n",
       "\n",
       "    virus_batch sample_type  antibody  \\\n",
       "48            1    antibody  AUSAB-07   \n",
       "49            1    antibody  AUSAB-07   \n",
       "50            1    antibody  AUSAB-07   \n",
       "51            1    antibody  AUSAB-07   \n",
       "60            1    antibody  AUSAB-07   \n",
       "\n",
       "                                    library_sample      condition  \n",
       "48  libA_220224_1_antibody_AUSAB-07_0.000158_1_rt1  serum-IC65.00  \n",
       "49  libA_220224_1_antibody_AUSAB-07_0.000158_1_rt1  serum-IC65.00  \n",
       "50  libA_220224_1_antibody_AUSAB-07_0.000158_1_rt1  serum-IC65.00  \n",
       "51  libA_220224_1_antibody_AUSAB-07_0.000158_1_rt1  serum-IC65.00  \n",
       "60      libA_220224_1_antibody_AUSAB-07_0.000395_1  serum-IC90.00  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "avg_muts_reformatted.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c099d6b-4499-427b-a5c2-9942f422dbca",
   "metadata": {},
   "source": [
    "Make an interactive chart of average mutations per variant.\n",
    "You can select which groupings to show, and click on mutation type to only show that mutation type:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2423eb9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>sample</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>count</th>\n",
       "      <th>number</th>\n",
       "      <th>date</th>\n",
       "      <th>virus_batch</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>antibody</th>\n",
       "      <th>library_sample</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>libA</td>\n",
       "      <td>no-Ab_1</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>2452009</td>\n",
       "      <td>2.739215</td>\n",
       "      <td>220224</td>\n",
       "      <td>1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "      <td>no antibody</td>\n",
       "      <td>libA_220224_1_no-antibody_control_670ng_1</td>\n",
       "      <td>no-Ab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>libA</td>\n",
       "      <td>no-Ab_1</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>2452009</td>\n",
       "      <td>0.173318</td>\n",
       "      <td>220224</td>\n",
       "      <td>1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "      <td>no antibody</td>\n",
       "      <td>libA_220224_1_no-antibody_control_670ng_1</td>\n",
       "      <td>no-Ab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>libA</td>\n",
       "      <td>no-Ab_1</td>\n",
       "      <td>stop</td>\n",
       "      <td>2452009</td>\n",
       "      <td>0.010307</td>\n",
       "      <td>220224</td>\n",
       "      <td>1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "      <td>no antibody</td>\n",
       "      <td>libA_220224_1_no-antibody_control_670ng_1</td>\n",
       "      <td>no-Ab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>libA</td>\n",
       "      <td>no-Ab_1</td>\n",
       "      <td>deletion</td>\n",
       "      <td>2452009</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>220224</td>\n",
       "      <td>1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "      <td>no antibody</td>\n",
       "      <td>libA_220224_1_no-antibody_control_670ng_1</td>\n",
       "      <td>no-Ab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>libA</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>9795560</td>\n",
       "      <td>3.449037</td>\n",
       "      <td>220224</td>\n",
       "      <td>1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "      <td>no antibody</td>\n",
       "      <td>libA_220224_1_no-antibody_control_plasmid</td>\n",
       "      <td>plasmid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>libA</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>9795560</td>\n",
       "      <td>0.190656</td>\n",
       "      <td>220224</td>\n",
       "      <td>1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "      <td>no antibody</td>\n",
       "      <td>libA_220224_1_no-antibody_control_plasmid</td>\n",
       "      <td>plasmid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>libA</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>stop</td>\n",
       "      <td>9795560</td>\n",
       "      <td>0.050777</td>\n",
       "      <td>220224</td>\n",
       "      <td>1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "      <td>no antibody</td>\n",
       "      <td>libA_220224_1_no-antibody_control_plasmid</td>\n",
       "      <td>plasmid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>libA</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>deletion</td>\n",
       "      <td>9795560</td>\n",
       "      <td>0.000459</td>\n",
       "      <td>220224</td>\n",
       "      <td>1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "      <td>no antibody</td>\n",
       "      <td>libA_220224_1_no-antibody_control_plasmid</td>\n",
       "      <td>plasmid</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    library   sample  mutation_type    count    number    date  virus_batch  \\\n",
       "108    libA  no-Ab_1  nonsynonymous  2452009  2.739215  220224            1   \n",
       "109    libA  no-Ab_1     synonymous  2452009  0.173318  220224            1   \n",
       "110    libA  no-Ab_1           stop  2452009  0.010307  220224            1   \n",
       "111    libA  no-Ab_1       deletion  2452009  0.000049  220224            1   \n",
       "116    libA  plasmid  nonsynonymous  9795560  3.449037  220224            1   \n",
       "117    libA  plasmid     synonymous  9795560  0.190656  220224            1   \n",
       "118    libA  plasmid           stop  9795560  0.050777  220224            1   \n",
       "119    libA  plasmid       deletion  9795560  0.000459  220224            1   \n",
       "\n",
       "             sample_type     antibody  \\\n",
       "108  no-antibody_control  no antibody   \n",
       "109  no-antibody_control  no antibody   \n",
       "110  no-antibody_control  no antibody   \n",
       "111  no-antibody_control  no antibody   \n",
       "116  no-antibody_control  no antibody   \n",
       "117  no-antibody_control  no antibody   \n",
       "118  no-antibody_control  no antibody   \n",
       "119  no-antibody_control  no antibody   \n",
       "\n",
       "                                library_sample condition  \n",
       "108  libA_220224_1_no-antibody_control_670ng_1     no-Ab  \n",
       "109  libA_220224_1_no-antibody_control_670ng_1     no-Ab  \n",
       "110  libA_220224_1_no-antibody_control_670ng_1     no-Ab  \n",
       "111  libA_220224_1_no-antibody_control_670ng_1     no-Ab  \n",
       "116  libA_220224_1_no-antibody_control_plasmid   plasmid  \n",
       "117  libA_220224_1_no-antibody_control_plasmid   plasmid  \n",
       "118  libA_220224_1_no-antibody_control_plasmid   plasmid  \n",
       "119  libA_220224_1_no-antibody_control_plasmid   plasmid  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "small_df = avg_muts_reformatted.loc[((avg_muts_reformatted['sample'] == 'plasmid') | (avg_muts_reformatted['sample'] == 'no-Ab_1'))]\n",
    "small_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "518ddfa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mut_types_sort = [\"nonsynonymous\", \"synonymous\", \"deletion\", \"stop\"]\n",
    "\n",
    "# selections = [\n",
    "#     alt.selection_single(\n",
    "#         fields=[col],\n",
    "#         bind=alt.binding_select(\n",
    "#             options=[None] + avg_muts_reformatted[col].unique().tolist(),\n",
    "#             labels=[\"all\"] + [str(x) for x in avg_muts_reformatted[col].unique()],\n",
    "#             name=col,\n",
    "#         )\n",
    "#     )\n",
    "#     for col in [\"library\", \"date\", \"sample_type\", \"virus_batch\", \"antibody\"]\n",
    "# ]\n",
    "\n",
    "mutation_type_selection = alt.selection_multi(fields=[\"mutation_type\"], bind=\"legend\")\n",
    "\n",
    "avg_muts_chart = (\n",
    "    alt.Chart(\n",
    "        avg_muts_reformatted.assign(order=lambda x: x[\"mutation_type\"]\n",
    "                        .map(lambda m: mut_types_sort.index(m))\n",
    "                        ),\n",
    "    )\n",
    "    .encode(\n",
    "        x=alt.X(\n",
    "            \"number\",\n",
    "            title=\"average mutations per variant\",\n",
    "            axis=alt.Axis(grid=False),\n",
    "        ),\n",
    "        y=alt.Y(\n",
    "            \"sample\",\n",
    "            title=None,\n",
    "            sort=['plasmid', 'no-Ab']\n",
    "        ),\n",
    "        color=alt.Color(\n",
    "            \"mutation_type\",\n",
    "            scale=alt.Scale(domain=mut_types_sort),\n",
    "        ),\n",
    "        order=alt.Order(\"order\", sort=\"descending\"),\n",
    "        tooltip=[alt.Tooltip(c, format=\".3g\") if c in [\"count\", \"number\"] else c\n",
    "                 for c in avg_muts.columns if c != \"sample\"],\n",
    "    )\n",
    "    .mark_bar()\n",
    "    .properties(height=alt.Step(13), width=300)\n",
    "    .configure_axis(labelLimit=500)\n",
    "    .add_selection(mutation_type_selection, *selections)\n",
    "    .transform_filter(mutation_type_selection)\n",
    ")\n",
    "\n",
    "for selection in selections:\n",
    "    avg_muts_chart = avg_muts_chart.transform_filter(selection)\n",
    "\n",
    "avg_muts_chart\n",
    "avg_muts_chart.save('scratch_notebooks/scratch_figs/muts_per_variant.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e603075b-c3f9-4fec-8cbf-c5a0103e8aa1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-b0bc45c4a0a04484b0bbc8c2f0434c44\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-b0bc45c4a0a04484b0bbc8c2f0434c44\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-b0bc45c4a0a04484b0bbc8c2f0434c44\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"4.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"axis\": {\"labelLimit\": 500}}, \"data\": {\"name\": \"data-c8e670ed7520b067e114cbe5a4b6e776\"}, \"mark\": \"bar\", \"encoding\": {\"color\": {\"field\": \"mutation_type\", \"scale\": {\"domain\": [\"nonsynonymous\", \"synonymous\", \"deletion\", \"stop\"]}, \"type\": \"nominal\"}, \"order\": {\"field\": \"order\", \"sort\": \"descending\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"library\", \"type\": \"nominal\"}, {\"field\": \"mutation_type\", \"type\": \"nominal\"}, {\"field\": \"count\", \"format\": \".3g\", \"type\": \"quantitative\"}, {\"field\": \"number\", \"format\": \".3g\", \"type\": \"quantitative\"}, {\"field\": \"date\", \"type\": \"quantitative\"}, {\"field\": \"virus_batch\", \"type\": \"quantitative\"}, {\"field\": \"sample_type\", \"type\": \"nominal\"}, {\"field\": \"antibody\", \"type\": \"nominal\"}, {\"field\": \"library_sample\", \"type\": \"nominal\"}], \"x\": {\"axis\": {\"grid\": false}, \"field\": \"number\", \"title\": \"average mutations per variant\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"sample\", \"title\": null, \"type\": \"nominal\"}}, \"height\": {\"step\": 13}, \"selection\": {\"selector017\": {\"type\": \"multi\", \"fields\": [\"mutation_type\"], \"bind\": \"legend\"}, \"selector012\": {\"type\": \"single\", \"fields\": [\"library\"], \"bind\": {\"input\": \"select\", \"options\": [null, \"libA\"], \"labels\": [\"all\", \"libA\"], \"name\": \"library\"}}, \"selector013\": {\"type\": \"single\", \"fields\": [\"date\"], \"bind\": {\"input\": \"select\", \"options\": [null, 220224], \"labels\": [\"all\", \"220224\"], \"name\": \"date\"}}, \"selector014\": {\"type\": \"single\", \"fields\": [\"sample_type\"], \"bind\": {\"input\": \"select\", \"options\": [null, \"antibody\", \"no-antibody_control\"], \"labels\": [\"all\", \"antibody\", \"no-antibody_control\"], \"name\": \"sample_type\"}}, \"selector015\": {\"type\": \"single\", \"fields\": [\"virus_batch\"], \"bind\": {\"input\": \"select\", \"options\": [null, 1], \"labels\": [\"all\", \"1\"], \"name\": \"virus_batch\"}}, \"selector016\": {\"type\": \"single\", \"fields\": [\"antibody\"], \"bind\": {\"input\": \"select\", \"options\": [null, \"AUSAB-07\", \"no antibody\"], \"labels\": [\"all\", \"AUSAB-07\", \"no antibody\"], \"name\": \"antibody\"}}}, \"transform\": [{\"filter\": {\"selection\": \"selector017\"}}, {\"filter\": {\"selection\": \"selector012\"}}, {\"filter\": {\"selection\": \"selector013\"}}, {\"filter\": {\"selection\": \"selector014\"}}, {\"filter\": {\"selection\": \"selector015\"}}, {\"filter\": {\"selection\": \"selector016\"}}], \"width\": 300, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"data-c8e670ed7520b067e114cbe5a4b6e776\": [{\"library\": \"libA\", \"sample\": \"serum-IC65_1_RT-1\", \"mutation_type\": \"nonsynonymous\", \"count\": 1892557, \"number\": 2.7224934308451476, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000158_1_rt1\", \"condition\": \"serum-IC65\", \"order\": 0}, {\"library\": \"libA\", \"sample\": \"serum-IC65_1_RT-1\", \"mutation_type\": \"synonymous\", \"count\": 1892557, \"number\": 0.1687515884594229, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000158_1_rt1\", \"condition\": \"serum-IC65\", \"order\": 1}, {\"library\": \"libA\", \"sample\": \"serum-IC65_1_RT-1\", \"mutation_type\": \"stop\", \"count\": 1892557, \"number\": 0.010323599236377028, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000158_1_rt1\", \"condition\": \"serum-IC65\", \"order\": 3}, {\"library\": \"libA\", \"sample\": \"serum-IC65_1_RT-1\", \"mutation_type\": \"deletion\", \"count\": 1892557, \"number\": 3.117475457806555e-05, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000158_1_rt1\", \"condition\": \"serum-IC65\", \"order\": 2}, {\"library\": \"libA\", \"sample\": \"serum-IC65_1_RT-2\", \"mutation_type\": \"nonsynonymous\", \"count\": 2193070, \"number\": 2.7166994213590994, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000158_1_rt2\", \"condition\": \"serum-IC65\", \"order\": 0}, {\"library\": \"libA\", \"sample\": \"serum-IC65_1_RT-2\", \"mutation_type\": \"synonymous\", \"count\": 2193070, \"number\": 0.17025174755023778, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000158_1_rt2\", \"condition\": \"serum-IC65\", \"order\": 1}, {\"library\": \"libA\", \"sample\": \"serum-IC65_1_RT-2\", \"mutation_type\": \"stop\", \"count\": 2193070, \"number\": 0.010359906432535213, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000158_1_rt2\", \"condition\": \"serum-IC65\", \"order\": 3}, {\"library\": \"libA\", \"sample\": \"serum-IC65_1_RT-2\", \"mutation_type\": \"deletion\", \"count\": 2193070, \"number\": 2.8270871426812643e-05, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000158_1_rt2\", \"condition\": \"serum-IC65\", \"order\": 2}, {\"library\": \"libA\", \"sample\": \"serum-IC65_2\", \"mutation_type\": \"nonsynonymous\", \"count\": 1713538, \"number\": 2.7155096647987964, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000158_2\", \"condition\": \"serum-IC65\", \"order\": 0}, {\"library\": \"libA\", \"sample\": \"serum-IC65_2\", \"mutation_type\": \"synonymous\", \"count\": 1713538, \"number\": 0.17061424958185928, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000158_2\", \"condition\": \"serum-IC65\", \"order\": 1}, {\"library\": \"libA\", \"sample\": \"serum-IC65_2\", \"mutation_type\": \"stop\", \"count\": 1713538, \"number\": 0.010216289338199678, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000158_2\", \"condition\": \"serum-IC65\", \"order\": 3}, {\"library\": \"libA\", \"sample\": \"serum-IC65_2\", \"mutation_type\": \"deletion\", \"count\": 1713538, \"number\": 2.3927102871369063e-05, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000158_2\", \"condition\": \"serum-IC65\", \"order\": 2}, {\"library\": \"libA\", \"sample\": \"serum-IC90_1\", \"mutation_type\": \"nonsynonymous\", \"count\": 2552542, \"number\": 2.712283284662897, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000395_1\", \"condition\": \"serum-IC90\", \"order\": 0}, {\"library\": \"libA\", \"sample\": \"serum-IC90_1\", \"mutation_type\": \"synonymous\", \"count\": 2552542, \"number\": 0.16849164479957626, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000395_1\", \"condition\": \"serum-IC90\", \"order\": 1}, {\"library\": \"libA\", \"sample\": \"serum-IC90_1\", \"mutation_type\": \"stop\", \"count\": 2552542, \"number\": 0.010497378691516143, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000395_1\", \"condition\": \"serum-IC90\", \"order\": 3}, {\"library\": \"libA\", \"sample\": \"serum-IC90_1\", \"mutation_type\": \"deletion\", \"count\": 2552542, \"number\": 3.173307236472505e-05, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.000395_1\", \"condition\": \"serum-IC90\", \"order\": 2}, {\"library\": \"libA\", \"sample\": \"serum-IC99_1\", \"mutation_type\": \"nonsynonymous\", \"count\": 2091776, \"number\": 2.686220226257496, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.00158_1\", \"condition\": \"serum-IC99\", \"order\": 0}, {\"library\": \"libA\", \"sample\": \"serum-IC99_1\", \"mutation_type\": \"synonymous\", \"count\": 2091776, \"number\": 0.16287403622567617, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.00158_1\", \"condition\": \"serum-IC99\", \"order\": 1}, {\"library\": \"libA\", \"sample\": \"serum-IC99_1\", \"mutation_type\": \"stop\", \"count\": 2091776, \"number\": 0.011002134071717047, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.00158_1\", \"condition\": \"serum-IC99\", \"order\": 3}, {\"library\": \"libA\", \"sample\": \"serum-IC99_1\", \"mutation_type\": \"deletion\", \"count\": 2091776, \"number\": 2.7249571655856078e-05, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.00158_1\", \"condition\": \"serum-IC99\", \"order\": 2}, {\"library\": \"libA\", \"sample\": \"serum-IC99.90_1\", \"mutation_type\": \"nonsynonymous\", \"count\": 1694145, \"number\": 2.7167562398732104, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.00632_1\", \"condition\": \"serum-IC99.90\", \"order\": 0}, {\"library\": \"libA\", \"sample\": \"serum-IC99.90_1\", \"mutation_type\": \"synonymous\", \"count\": 1694145, \"number\": 0.15905781382349207, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.00632_1\", \"condition\": \"serum-IC99.90\", \"order\": 1}, {\"library\": \"libA\", \"sample\": \"serum-IC99.90_1\", \"mutation_type\": \"stop\", \"count\": 1694145, \"number\": 0.012552644549315437, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.00632_1\", \"condition\": \"serum-IC99.90\", \"order\": 3}, {\"library\": \"libA\", \"sample\": \"serum-IC99.90_1\", \"mutation_type\": \"deletion\", \"count\": 1694145, \"number\": 4.604092329759259e-05, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.00632_1\", \"condition\": \"serum-IC99.90\", \"order\": 2}, {\"library\": \"libA\", \"sample\": \"serum-IC99.90_2\", \"mutation_type\": \"nonsynonymous\", \"count\": 1865542, \"number\": 2.725025756589774, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.00632_2\", \"condition\": \"serum-IC99.90\", \"order\": 0}, {\"library\": \"libA\", \"sample\": \"serum-IC99.90_2\", \"mutation_type\": \"synonymous\", \"count\": 1865542, \"number\": 0.16018454690379524, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.00632_2\", \"condition\": \"serum-IC99.90\", \"order\": 1}, {\"library\": \"libA\", \"sample\": \"serum-IC99.90_2\", \"mutation_type\": \"stop\", \"count\": 1865542, \"number\": 0.012201279842533698, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.00632_2\", \"condition\": \"serum-IC99.90\", \"order\": 3}, {\"library\": \"libA\", \"sample\": \"serum-IC99.90_2\", \"mutation_type\": \"deletion\", \"count\": 1865542, \"number\": 2.7337899655971295e-05, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.00632_2\", \"condition\": \"serum-IC99.90\", \"order\": 2}, {\"library\": \"libA\", \"sample\": \"serum-IC99.99_1\", \"mutation_type\": \"nonsynonymous\", \"count\": 2237126, \"number\": 3.1734734655088714, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.0158_1\", \"condition\": \"serum-IC99.99\", \"order\": 0}, {\"library\": \"libA\", \"sample\": \"serum-IC99.99_1\", \"mutation_type\": \"synonymous\", \"count\": 2237126, \"number\": 0.16712827082605092, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.0158_1\", \"condition\": \"serum-IC99.99\", \"order\": 1}, {\"library\": \"libA\", \"sample\": \"serum-IC99.99_1\", \"mutation_type\": \"stop\", \"count\": 2237126, \"number\": 0.013211593803835815, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.0158_1\", \"condition\": \"serum-IC99.99\", \"order\": 3}, {\"library\": \"libA\", \"sample\": \"serum-IC99.99_1\", \"mutation_type\": \"deletion\", \"count\": 2237126, \"number\": 6.9732326207822e-05, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.0158_1\", \"condition\": \"serum-IC99.99\", \"order\": 2}, {\"library\": \"libA\", \"sample\": \"serum-IC99.99_2_RT-1\", \"mutation_type\": \"nonsynonymous\", \"count\": 2033210, \"number\": 3.1207755224497222, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.0158_2_rt1\", \"condition\": \"serum-IC99.99\", \"order\": 0}, {\"library\": \"libA\", \"sample\": \"serum-IC99.99_2_RT-1\", \"mutation_type\": \"synonymous\", \"count\": 2033210, \"number\": 0.1671671888294864, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.0158_2_rt1\", \"condition\": \"serum-IC99.99\", \"order\": 1}, {\"library\": \"libA\", \"sample\": \"serum-IC99.99_2_RT-1\", \"mutation_type\": \"stop\", \"count\": 2033210, \"number\": 0.012448296044186286, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.0158_2_rt1\", \"condition\": \"serum-IC99.99\", \"order\": 3}, {\"library\": \"libA\", \"sample\": \"serum-IC99.99_2_RT-1\", \"mutation_type\": \"deletion\", \"count\": 2033210, \"number\": 6.639747001047605e-05, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.0158_2_rt1\", \"condition\": \"serum-IC99.99\", \"order\": 2}, {\"library\": \"libA\", \"sample\": \"serum-IC99.99_2_RT-2\", \"mutation_type\": \"nonsynonymous\", \"count\": 1851798, \"number\": 3.122428580223113, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.0158_2_rt2\", \"condition\": \"serum-IC99.99\", \"order\": 0}, {\"library\": \"libA\", \"sample\": \"serum-IC99.99_2_RT-2\", \"mutation_type\": \"synonymous\", \"count\": 1851798, \"number\": 0.16613367116715755, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.0158_2_rt2\", \"condition\": \"serum-IC99.99\", \"order\": 1}, {\"library\": \"libA\", \"sample\": \"serum-IC99.99_2_RT-2\", \"mutation_type\": \"stop\", \"count\": 1851798, \"number\": 0.012341518891369361, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.0158_2_rt2\", \"condition\": \"serum-IC99.99\", \"order\": 3}, {\"library\": \"libA\", \"sample\": \"serum-IC99.99_2_RT-2\", \"mutation_type\": \"deletion\", \"count\": 1851798, \"number\": 6.966202577170944e-05, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"antibody\", \"antibody\": \"AUSAB-07\", \"library_sample\": \"libA_220224_1_antibody_AUSAB-07_0.0158_2_rt2\", \"condition\": \"serum-IC99.99\", \"order\": 2}, {\"library\": \"libA\", \"sample\": \"no-Ab_1\", \"mutation_type\": \"nonsynonymous\", \"count\": 2452009, \"number\": 2.7392146603050804, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"no-antibody_control\", \"antibody\": \"no antibody\", \"library_sample\": \"libA_220224_1_no-antibody_control_670ng_1\", \"condition\": \"no-Ab\", \"order\": 0}, {\"library\": \"libA\", \"sample\": \"no-Ab_1\", \"mutation_type\": \"synonymous\", \"count\": 2452009, \"number\": 0.17331787933894208, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"no-antibody_control\", \"antibody\": \"no antibody\", \"library_sample\": \"libA_220224_1_no-antibody_control_670ng_1\", \"condition\": \"no-Ab\", \"order\": 1}, {\"library\": \"libA\", \"sample\": \"no-Ab_1\", \"mutation_type\": \"stop\", \"count\": 2452009, \"number\": 0.010307466245026018, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"no-antibody_control\", \"antibody\": \"no antibody\", \"library_sample\": \"libA_220224_1_no-antibody_control_670ng_1\", \"condition\": \"no-Ab\", \"order\": 3}, {\"library\": \"libA\", \"sample\": \"no-Ab_1\", \"mutation_type\": \"deletion\", \"count\": 2452009, \"number\": 4.853163263266978e-05, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"no-antibody_control\", \"antibody\": \"no antibody\", \"library_sample\": \"libA_220224_1_no-antibody_control_670ng_1\", \"condition\": \"no-Ab\", \"order\": 2}, {\"library\": \"libA\", \"sample\": \"no-Ab_2\", \"mutation_type\": \"nonsynonymous\", \"count\": 2380112, \"number\": 2.74083656567422, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"no-antibody_control\", \"antibody\": \"no antibody\", \"library_sample\": \"libA_220224_1_no-antibody_control_670ng_2\", \"condition\": \"no-Ab\", \"order\": 0}, {\"library\": \"libA\", \"sample\": \"no-Ab_2\", \"mutation_type\": \"synonymous\", \"count\": 2380112, \"number\": 0.17333470021578817, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"no-antibody_control\", \"antibody\": \"no antibody\", \"library_sample\": \"libA_220224_1_no-antibody_control_670ng_2\", \"condition\": \"no-Ab\", \"order\": 1}, {\"library\": \"libA\", \"sample\": \"no-Ab_2\", \"mutation_type\": \"stop\", \"count\": 2380112, \"number\": 0.010168008900421494, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"no-antibody_control\", \"antibody\": \"no antibody\", \"library_sample\": \"libA_220224_1_no-antibody_control_670ng_2\", \"condition\": \"no-Ab\", \"order\": 3}, {\"library\": \"libA\", \"sample\": \"no-Ab_2\", \"mutation_type\": \"deletion\", \"count\": 2380112, \"number\": 3.193127046122199e-05, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"no-antibody_control\", \"antibody\": \"no antibody\", \"library_sample\": \"libA_220224_1_no-antibody_control_670ng_2\", \"condition\": \"no-Ab\", \"order\": 2}, {\"library\": \"libA\", \"sample\": \"plasmid\", \"mutation_type\": \"nonsynonymous\", \"count\": 9795560, \"number\": 3.4490374210356527, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"no-antibody_control\", \"antibody\": \"no antibody\", \"library_sample\": \"libA_220224_1_no-antibody_control_plasmid\", \"condition\": \"plasmid\", \"order\": 0}, {\"library\": \"libA\", \"sample\": \"plasmid\", \"mutation_type\": \"synonymous\", \"count\": 9795560, \"number\": 0.1906556644030561, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"no-antibody_control\", \"antibody\": \"no antibody\", \"library_sample\": \"libA_220224_1_no-antibody_control_plasmid\", \"condition\": \"plasmid\", \"order\": 1}, {\"library\": \"libA\", \"sample\": \"plasmid\", \"mutation_type\": \"stop\", \"count\": 9795560, \"number\": 0.05077657632641727, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"no-antibody_control\", \"antibody\": \"no antibody\", \"library_sample\": \"libA_220224_1_no-antibody_control_plasmid\", \"condition\": \"plasmid\", \"order\": 3}, {\"library\": \"libA\", \"sample\": \"plasmid\", \"mutation_type\": \"deletion\", \"count\": 9795560, \"number\": 0.00045908554487951686, \"date\": 220224, \"virus_batch\": 1, \"sample_type\": \"no-antibody_control\", \"antibody\": \"no antibody\", \"library_sample\": \"libA_220224_1_no-antibody_control_plasmid\", \"condition\": \"plasmid\", \"order\": 2}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_types_sort = [\"nonsynonymous\", \"synonymous\", \"deletion\", \"stop\"]\n",
    "\n",
    "selections = [\n",
    "    alt.selection_single(\n",
    "        fields=[col],\n",
    "        bind=alt.binding_select(\n",
    "            options=[None] + avg_muts_reformatted[col].unique().tolist(),\n",
    "            labels=[\"all\"] + [str(x) for x in avg_muts_reformatted[col].unique()],\n",
    "            name=col,\n",
    "        )\n",
    "    )\n",
    "    for col in [\"library\", \"date\", \"sample_type\", \"virus_batch\", \"antibody\"]\n",
    "]\n",
    "\n",
    "mutation_type_selection = alt.selection_multi(fields=[\"mutation_type\"], bind=\"legend\")\n",
    "\n",
    "avg_muts_chart = (\n",
    "    alt.Chart(\n",
    "        avg_muts_reformatted.assign(order=lambda x: x[\"mutation_type\"]\n",
    "                        .map(lambda m: mut_types_sort.index(m))\n",
    "                        ),\n",
    "    )\n",
    "    .encode(\n",
    "        x=alt.X(\n",
    "            \"number\",\n",
    "            title=\"average mutations per variant\",\n",
    "            axis=alt.Axis(grid=False),\n",
    "        ),\n",
    "        y=alt.Y(\n",
    "            \"sample\",\n",
    "            title=None,\n",
    "        ),\n",
    "        color=alt.Color(\n",
    "            \"mutation_type\",\n",
    "            scale=alt.Scale(domain=mut_types_sort),\n",
    "        ),\n",
    "        order=alt.Order(\"order\", sort=\"descending\"),\n",
    "        tooltip=[alt.Tooltip(c, format=\".3g\") if c in [\"count\", \"number\"] else c\n",
    "                 for c in avg_muts.columns if c != \"sample\"],\n",
    "    )\n",
    "    .mark_bar()\n",
    "    .properties(height=alt.Step(13), width=300)\n",
    "    .configure_axis(labelLimit=500)\n",
    "    .add_selection(mutation_type_selection, *selections)\n",
    "    .transform_filter(mutation_type_selection)\n",
    ")\n",
    "\n",
    "for selection in selections:\n",
    "    avg_muts_chart = avg_muts_chart.transform_filter(selection)\n",
    "\n",
    "avg_muts_chart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95ffc9fa-a9fb-413b-8628-8141d49d04c0",
   "metadata": {},
   "source": [
    "## Mutation frequencies along the gene\n",
    "Get counts of amino-acid mutations per-site along the gene:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "804e7fe2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>sample</th>\n",
       "      <th>mutation</th>\n",
       "      <th>count</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>libA</td>\n",
       "      <td>220121_1_no-antibody_control_100x_1</td>\n",
       "      <td>K297I</td>\n",
       "      <td>208154</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>libA</td>\n",
       "      <td>220121_1_no-antibody_control_100x_1</td>\n",
       "      <td>L263H</td>\n",
       "      <td>156635</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>libA</td>\n",
       "      <td>220121_1_no-antibody_control_100x_1</td>\n",
       "      <td>L263S</td>\n",
       "      <td>113951</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>libA</td>\n",
       "      <td>220121_1_no-antibody_control_100x_1</td>\n",
       "      <td>V366M</td>\n",
       "      <td>81112</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>libA</td>\n",
       "      <td>220121_1_no-antibody_control_100x_1</td>\n",
       "      <td>R402S</td>\n",
       "      <td>76351</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359095</th>\n",
       "      <td>libA</td>\n",
       "      <td>220224_1_no-antibody_control_plasmid</td>\n",
       "      <td>Y530R</td>\n",
       "      <td>0</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359096</th>\n",
       "      <td>libA</td>\n",
       "      <td>220224_1_no-antibody_control_plasmid</td>\n",
       "      <td>Y530S</td>\n",
       "      <td>0</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359097</th>\n",
       "      <td>libA</td>\n",
       "      <td>220224_1_no-antibody_control_plasmid</td>\n",
       "      <td>Y530T</td>\n",
       "      <td>0</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359098</th>\n",
       "      <td>libA</td>\n",
       "      <td>220224_1_no-antibody_control_plasmid</td>\n",
       "      <td>Y530V</td>\n",
       "      <td>0</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359099</th>\n",
       "      <td>libA</td>\n",
       "      <td>220224_1_no-antibody_control_plasmid</td>\n",
       "      <td>Y530W</td>\n",
       "      <td>0</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>530</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>359100 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       library                                sample mutation   count  \\\n",
       "0         libA   220121_1_no-antibody_control_100x_1    K297I  208154   \n",
       "1         libA   220121_1_no-antibody_control_100x_1    L263H  156635   \n",
       "2         libA   220121_1_no-antibody_control_100x_1    L263S  113951   \n",
       "3         libA   220121_1_no-antibody_control_100x_1    V366M   81112   \n",
       "4         libA   220121_1_no-antibody_control_100x_1    R402S   76351   \n",
       "...        ...                                   ...      ...     ...   \n",
       "359095    libA  220224_1_no-antibody_control_plasmid    Y530R       0   \n",
       "359096    libA  220224_1_no-antibody_control_plasmid    Y530S       0   \n",
       "359097    libA  220224_1_no-antibody_control_plasmid    Y530T       0   \n",
       "359098    libA  220224_1_no-antibody_control_plasmid    Y530V       0   \n",
       "359099    libA  220224_1_no-antibody_control_plasmid    Y530W       0   \n",
       "\n",
       "        mutation_type  site  \n",
       "0       nonsynonymous   297  \n",
       "1       nonsynonymous   263  \n",
       "2       nonsynonymous   263  \n",
       "3       nonsynonymous   366  \n",
       "4       nonsynonymous   402  \n",
       "...               ...   ...  \n",
       "359095  nonsynonymous   530  \n",
       "359096  nonsynonymous   530  \n",
       "359097  nonsynonymous   530  \n",
       "359098  nonsynonymous   530  \n",
       "359099  nonsynonymous   530  \n",
       "\n",
       "[359100 rows x 6 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "    variants.mutCounts(\n",
    "        variant_type=\"all\",\n",
    "        mut_type=\"aa\",\n",
    "        libraries=variants.libraries,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "3cded3c8-1375-459d-b867-85a9f372d55f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# read reference site numbering\n",
    "site_numbering_map = pd.read_csv(config[\"site_numbering_map\"])\n",
    "\n",
    "site_freqs = (\n",
    "    variants.mutCounts(\n",
    "        variant_type=\"all\",\n",
    "        mut_type=\"aa\",\n",
    "        libraries=variants.libraries,\n",
    "    )\n",
    "    .query(\"count > 0\")\n",
    "    .rename(columns={\"site\": \"sequential_site\"})\n",
    "    .merge(\n",
    "        site_numbering_map,\n",
    "        how=\"left\",\n",
    "        on=\"sequential_site\",\n",
    "        validate=\"many_to_one\",\n",
    "    )\n",
    "    .assign(\n",
    "        wildtype=lambda x: x[\"mutation\"].str[0],\n",
    "        mutant=lambda x: x[\"mutation\"].str[-1],\n",
    "        reference_site=lambda x: x[\"reference_site\"].astype(str),\n",
    "    )\n",
    "    .assign(mutant=lambda x: x[\"mutant\"] + \"=\" + x[\"count\"].map(\"{:.2g}\".format))\n",
    "    .sort_values(\"count\", ascending=False)\n",
    "    .groupby(\n",
    "        [\"library\", \"sample\", \"sequential_site\", \"reference_site\", \"wildtype\"],\n",
    "        observed=True,\n",
    "        as_index=False,\n",
    "    )\n",
    "    .aggregate(\n",
    "        count=pd.NamedAgg(\"count\", \"sum\"),\n",
    "        mutants=pd.NamedAgg(\"mutant\", \", \".join),\n",
    "    )\n",
    "    .merge(\n",
    "        variants.n_variants_df(libraries=variants.libraries, primary_target_only=True)\n",
    "        .rename(columns={\"count\": \"n_variants\"})\n",
    "    )\n",
    "    .assign(percent=lambda x: 100 * x[\"count\"] / x[\"n_variants\"])\n",
    "    .merge(barcode_runs[[\"library\", \"sample\", \"library_sample\", \"sample_type\"]])\n",
    "    .drop(columns=[\"library\", \"sample\"])\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "1923d8e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequential_site</th>\n",
       "      <th>reference_site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>count</th>\n",
       "      <th>mutants</th>\n",
       "      <th>n_variants</th>\n",
       "      <th>percent</th>\n",
       "      <th>library_sample</th>\n",
       "      <th>sample_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>-19</td>\n",
       "      <td>M</td>\n",
       "      <td>3</td>\n",
       "      <td>I=3</td>\n",
       "      <td>7789129</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>libA_220121_1_no-antibody_control_100x_1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>-17</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>T=4</td>\n",
       "      <td>7789129</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>libA_220121_1_no-antibody_control_100x_1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>-9</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>S=2</td>\n",
       "      <td>7789129</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>libA_220121_1_no-antibody_control_100x_1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>-7</td>\n",
       "      <td>V</td>\n",
       "      <td>9</td>\n",
       "      <td>L=9</td>\n",
       "      <td>7789129</td>\n",
       "      <td>0.000116</td>\n",
       "      <td>libA_220121_1_no-antibody_control_100x_1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14</td>\n",
       "      <td>-6</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>S=6</td>\n",
       "      <td>7789129</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>libA_220121_1_no-antibody_control_100x_1</td>\n",
       "      <td>no-antibody_control</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sequential_site reference_site wildtype  count mutants  n_variants  \\\n",
       "0                1            -19        M      3     I=3     7789129   \n",
       "1                3            -17        A      4     T=4     7789129   \n",
       "2               11             -9        A      2     S=2     7789129   \n",
       "3               13             -7        V      9     L=9     7789129   \n",
       "4               14             -6        A      6     S=6     7789129   \n",
       "\n",
       "    percent                            library_sample          sample_type  \n",
       "0  0.000039  libA_220121_1_no-antibody_control_100x_1  no-antibody_control  \n",
       "1  0.000051  libA_220121_1_no-antibody_control_100x_1  no-antibody_control  \n",
       "2  0.000026  libA_220121_1_no-antibody_control_100x_1  no-antibody_control  \n",
       "3  0.000116  libA_220121_1_no-antibody_control_100x_1  no-antibody_control  \n",
       "4  0.000077  libA_220121_1_no-antibody_control_100x_1  no-antibody_control  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "site_freqs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "d1b8d35a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# edit names for virus meeting, drop to just 220224 samples\n",
    "\n",
    "# avg_muts_reformatted = avg_muts.duplicate()\n",
    "\n",
    "short_names = {\n",
    "#     '220121_1_no-antibody_control_50x_1': 'no-Ab_1',\n",
    "#     '220121_1_no-antibody_control_50x_2_rt1': 'no-Ab_2_RT-1',\n",
    "#     '220121_1_no-antibody_control_50x_2_rt2': 'no-Ab_2_RT-2',\n",
    "    'libA_220224_1_antibody_AUSAB-07_0.000158_1_rt1': 'AUSAB-07_IC65.00',\n",
    "#     '220224_1_antibody_AUSAB-07_0.000158_1_rt2': 'serum-IC65_1_RT-2',\n",
    "#     '220224_1_antibody_AUSAB-07_0.000158_2': 'serum-IC65_2',\n",
    "    'libA_220224_1_antibody_AUSAB-07_0.000395_1': 'AUSAB-07_IC90.00',\n",
    "#     '220224_1_antibody_AUSAB-07_0.000395_2': 'serum-IC90_2',\n",
    "    'libA_220224_1_antibody_AUSAB-07_0.00158_1': 'AUSAB-07_IC99.00',\n",
    "#     '220224_1_antibody_AUSAB-07_0.00158_2': 'serum-IC99_2',\n",
    "    'libA_220224_1_antibody_AUSAB-07_0.00632_1': 'AUSAB-07_IC99.90',\n",
    "#     '220224_1_antibody_AUSAB-07_0.00632_2': 'serum-IC99.90_2',\n",
    "    'libA_220224_1_antibody_AUSAB-07_0.0158_1': 'AUSAB-07_IC99.99',\n",
    "#     '220224_1_antibody_AUSAB-07_0.0158_2_rt1': 'serum-IC99.99_2_RT-1',\n",
    "#     '220224_1_antibody_AUSAB-07_0.0158_2_rt2': 'serum-IC99.99_2_RT-2',\n",
    "    'libA_220224_1_no-antibody_control_670ng_1': 'no-Ab_infection',\n",
    "#     '220224_1_no-antibody_control_670ng_2': 'no-Ab_2',\n",
    "    'libA_220224_1_no-antibody_control_plasmid': 'plasmid'\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "site_freqs['library_sample'] = site_freqs['library_sample'].map(short_names)\n",
    "site_freqs = site_freqs.dropna(subset=['library_sample'])\n",
    "\n",
    "# d['condition'] = avg_muts_reformatted['sample'].str.split('_').str[0]  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7b6c9be-f817-4954-8974-29a2cf19ec8b",
   "metadata": {},
   "source": [
    "Now make chart.\n",
    "You can mouse over points to see details.\n",
    "Note sites are numbered on the x-axis in sequential numbering, but the mouseovers also show the reference-based numbering:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "5fa2cdd8-db3c-490e-8f00-434e4d747bb1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "zoom_brush = alt.selection_interval(\n",
    "    encodings=['x'], mark=alt.BrushConfig(stroke='black', strokeWidth=2),\n",
    ")\n",
    "zoom_bar = (\n",
    "    alt.Chart(site_freqs[[\"sequential_site\"]].drop_duplicates())\n",
    "    .mark_rect(color='lightgrey')\n",
    "    .encode(x=alt.X(\"sequential_site\", title=None, scale=alt.Scale(nice=False, zero=False)))\n",
    "    .add_selection(zoom_brush)\n",
    "    .properties(width=500, height=15, title='site zoom bar')\n",
    ")\n",
    "\n",
    "site_freqs_base = (\n",
    "    alt.Chart()\n",
    "    .encode(\n",
    "        x=alt.X(\"sequential_site\", scale=alt.Scale(nice=False, zero=False)),\n",
    "        y=alt.Y('percent', title='% variants with mutation'),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if c in {\"percent\", \"count\", \"n_variants\"} else c\n",
    "            for c in site_freqs.columns\n",
    "        ],\n",
    "        color=\"sample_type\",\n",
    "    )\n",
    "    .properties(height=125, width=375)\n",
    ")\n",
    "\n",
    "site_freqs_chart = (\n",
    "    alt.layer(\n",
    "        site_freqs_base.mark_point(filled=True, size=25),\n",
    "        site_freqs_base.mark_line(size=0.4),\n",
    "        data=site_freqs,\n",
    "    )\n",
    "    .add_selection(zoom_brush)\n",
    "    .facet(facet=alt.Facet(\"library_sample\", title=None, sort=['plasmid', 'no-Ab_infection']), columns=2)\n",
    "    .transform_filter(zoom_brush)\n",
    ")\n",
    "    \n",
    "site_freqs_chart = site_freqs_chart\n",
    "\n",
    "site_freqs_zoom_chart = (zoom_bar & site_freqs_chart).configure_axis(grid=False)\n",
    "\n",
    "site_freqs_zoom_chart.save('scratch_notebooks/scratch_figs/AUSAB-07_selection.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e73b34d8-99a7-4f1f-9b93-04f5b3f43719",
   "metadata": {},
   "source": [
    "## To what extent do individual variants dominate counts?\n",
    "We look to see how much individual variants dominate the counts, only looking at the primary target (not any neutralization standards):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "273b6d57-99a4-46e1-9d4f-fd926746f0e0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "variant_counts = (\n",
    "    variants.variant_count_df\n",
    "    .query(\"target == @primary_target\")\n",
    "    [[\"library\", \"sample\", \"barcode\", \"count\", \"aa_substitutions\"]]\n",
    "    .merge(barcode_runs.drop(columns=[\"fastq_R1\", \"notes\", \"antibody_concentration\"]))\n",
    "    .assign(percent=lambda x: 100 * x[\"count\"] / x.groupby(\"library_sample\")[\"count\"].transform(\"sum\"))\n",
    "    .sort_values([\"library\", \"sample\", \"count\"], ascending=[True, True, False])\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3755a290-10df-4240-8c7d-6d45073dc00b",
   "metadata": {},
   "source": [
    "Get the top 25 variants, and the 10th and 90th percentiles:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bebe191c-740f-4445-a727-909689580e12",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "top_n = 25\n",
    "\n",
    "variant_counts_top_n = (\n",
    "    variant_counts\n",
    "    .groupby(\"library_sample\")\n",
    "    .head(n=top_n)\n",
    "    .merge(\n",
    "        (\n",
    "            variant_counts\n",
    "            .groupby(\"library_sample\", as_index=False)\n",
    "            .aggregate(\n",
    "                percentile_10=pd.NamedAgg(\"percent\", lambda s: s.quantile(0.1)),\n",
    "                percentile_90=pd.NamedAgg(\"percent\", lambda s: s.quantile(0.9)),\n",
    "                min_percent=pd.NamedAgg(\"percent\", \"min\"),\n",
    "                max_percent=pd.NamedAgg(\"percent\", \"max\"),\n",
    "            )\n",
    "        ),\n",
    "        validate=\"many_to_one\",\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41b210ef-f31c-4bdf-97f5-24df7b555863",
   "metadata": {},
   "source": [
    "Now plot these data.\n",
    "The points show the top variants, and can be moused over for details.\n",
    "The bars show the 10th to 90th percentiles, and the lines the full span.\n",
    "Note that these are the **percent** frequencies of variants, not the fractional frequencies:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96d89578-22e7-4856-848c-3254f763c5ae",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "variant_selector = alt.selection_single(\n",
    "    on=\"mouseover\",\n",
    "    empty=\"none\",\n",
    "    fields=[\"barcode\", \"library\"],\n",
    ")\n",
    "\n",
    "variant_counts_top_n_base = (\n",
    "    alt.Chart(variant_counts_top_n)\n",
    "    .encode(\n",
    "        y=alt.Y(\"library_sample\", title=None),\n",
    "        color=alt.Color(\n",
    "            \"sample_type\",\n",
    "            scale=alt.Scale(domain=variant_counts_top_n[\"sample_type\"].unique()),\n",
    "        ),\n",
    "    )\n",
    "    .mark_point(filled=True)\n",
    "    .properties(height=alt.Step(15), width=275)\n",
    ")\n",
    "\n",
    "variant_counts_top_n_points = (\n",
    "    variant_counts_top_n_base\n",
    "    .encode(\n",
    "        x=alt.X(\"percent\", title=\"percent of library\"),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3g\") if c.startswith(\"percent\") else c\n",
    "            for c in variant_counts_top_n.columns\n",
    "            if c not in [\"library_sample\", \"min_percent\", \"max_percent\"]\n",
    "        ],\n",
    "        opacity=alt.condition(variant_selector, alt.value(1), alt.value(0.5)),\n",
    "        strokeWidth=alt.condition(variant_selector, alt.value(2), alt.value(0)),\n",
    "        size=alt.condition(variant_selector, alt.value(40), alt.value(25)),\n",
    "    )\n",
    "    .mark_point(filled=True, stroke=\"black\")\n",
    ")\n",
    "\n",
    "variant_counts_top_n_bars = (\n",
    "    variant_counts_top_n_base\n",
    "    .encode(\n",
    "        alt.X(\"percentile_10\"),\n",
    "        alt.X2(\"percentile_90\"),\n",
    "    )\n",
    "    .mark_bar(size=11)\n",
    ")\n",
    "\n",
    "variant_counts_top_n_rule = (\n",
    "    variant_counts_top_n_base\n",
    "    .encode(\n",
    "        alt.X(\"min_percent\"),\n",
    "        alt.X2(\"max_percent\"),\n",
    "    )\n",
    "    .mark_rule()\n",
    ")\n",
    "\n",
    "variant_counts_top_n_chart = (\n",
    "    (variant_counts_top_n_points + variant_counts_top_n_bars + variant_counts_top_n_rule)\n",
    "    .configure_axis(labelLimit=500)\n",
    "    .add_selection(variant_selector, *selections)\n",
    ")\n",
    "for selection in selections:\n",
    "    variant_counts_top_n_chart = variant_counts_top_n_chart.transform_filter(selection)\n",
    "\n",
    "variant_counts_top_n_chart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24165dce-ce0f-4f74-9806-7be03d4a645e",
   "metadata": {},
   "source": [
    "## Pairwise correlation between variant counts\n",
    "Compute the pairwise correlations between variant counts for each library:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ede2944-e5ce-4458-842a-e827863041fe",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "corrs = (\n",
    "    dms_variants.utils.tidy_to_corr(\n",
    "        df=variant_counts,\n",
    "        sample_col=\"sample\",\n",
    "        label_col=\"barcode\",\n",
    "        value_col=\"count\",\n",
    "        group_cols=[\"library\"],\n",
    "    )\n",
    "    .assign(r2=lambda x: x[\"correlation\"]**2)\n",
    "    .drop(columns=\"correlation\")\n",
    "    .sort_values([\"library\", \"sample_1\", \"sample_2\"])\n",
    ")\n",
    "\n",
    "# add other properties and flag samples that share these\n",
    "suffixes = [\"_1\", \"_2\"]\n",
    "cols_to_add =  [\"antibody\", \"virus_batch\", \"sample_type\", \"date\"]\n",
    "for suffix in suffixes:\n",
    "    corrs = (\n",
    "        corrs.merge(\n",
    "            barcode_runs[[\"library\", \"sample\", *cols_to_add]],\n",
    "            left_on=[\"library\", f\"sample{suffix}\"],\n",
    "            right_on=[\"library\", \"sample\"],\n",
    "            validate=\"many_to_one\",\n",
    "            suffixes=suffixes,\n",
    "        )\n",
    "        .drop(columns=\"sample\")\n",
    "    )\n",
    "for col in cols_to_add:\n",
    "    corrs = (\n",
    "        corrs\n",
    "        .assign(\n",
    "            equal=lambda x: x[f\"{col}_1\"] == x[f\"{col}_2\"],\n",
    "            **{col: lambda x: x[f\"{col}_1\"].where(x[\"equal\"], pd.NA)},\n",
    "        )\n",
    "        .drop(columns=[f\"{col}{suffix}\" for suffix in suffixes])\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "921afbe8-2fcb-4bad-ada1-efe6fa8b206e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "corr_charts = []\n",
    "for library, library_corr in corrs.groupby(\"library\"):\n",
    "    corr_chart = (\n",
    "        alt.Chart(library_corr)\n",
    "        .encode(\n",
    "            alt.X(\"sample_1\", title=None),\n",
    "            alt.Y(\"sample_2\", title=None),\n",
    "            color=alt.Color(\"r2\", scale=alt.Scale(zero=True)),\n",
    "            tooltip=[\n",
    "                \"library\",\n",
    "                \"sample_1\",\n",
    "                \"sample_2\",\n",
    "                alt.Tooltip(\"r2\", format=\".3f\")\n",
    "            ],\n",
    "        )\n",
    "        .mark_rect(stroke=\"black\")\n",
    "        .properties(width=alt.Step(15), height=alt.Step(15), title=library)\n",
    "        .add_selection(*selections)\n",
    "    )\n",
    "    for selection in selections:\n",
    "        corr_chart = corr_chart.transform_filter(selection)\n",
    "    corr_charts.append(corr_chart)\n",
    "\n",
    "corr_chart = (\n",
    "    alt.vconcat(*corr_charts)\n",
    "    .configure_axis(labelLimit=500)\n",
    "\n",
    ")\n",
    "\n",
    "corr_chart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af6e8502",
   "metadata": {},
   "source": [
    "Try plotting on a log scale, to better resolve differences in pairwise correlations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b6b29d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "corr_charts = []\n",
    "for library, library_corr in corrs.groupby(\"library\"):\n",
    "    corr_chart = (\n",
    "        alt.Chart(library_corr)\n",
    "        .encode(\n",
    "            alt.X(\"sample_1\", title=None),\n",
    "            alt.Y(\"sample_2\", title=None),\n",
    "            color=alt.Color(\"r2\", scale=alt.Scale(type='log')),\n",
    "            tooltip=[\n",
    "                \"library\",\n",
    "                \"sample_1\",\n",
    "                \"sample_2\",\n",
    "                alt.Tooltip(\"r2\", format=\".3f\")\n",
    "            ],\n",
    "        )\n",
    "        .mark_rect(stroke=\"black\")\n",
    "        .properties(width=alt.Step(15), height=alt.Step(15), title=library)\n",
    "        .add_selection(*selections)\n",
    "    )\n",
    "    for selection in selections:\n",
    "        corr_chart = corr_chart.transform_filter(selection)\n",
    "    corr_charts.append(corr_chart)\n",
    "\n",
    "corr_chart = (\n",
    "    alt.vconcat(*corr_charts)\n",
    "    .configure_axis(labelLimit=500)\n",
    "\n",
    ")\n",
    "\n",
    "corr_chart"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
